doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,230 @@
+has been received, and any timestamp selected for a read will be susceptible to one of the negative scenarios described
+above. This means is that if any upstream source is unresponsive to an acknowledgement, then all reads will be halted
+until that source is responsive again.
+
+Property 1 alone is still a useful property, because property 1 allows us to provide guarantees like the following: A
+client is using synchronous replication with an upstream source. When their transaction commits in the upstream source,
+they are guaranteed to also be able to see the transaction in Materialize. However, they may actually see the
+transaction in Materialize before their transaction commits in the upstream source. This is still a powerful and useful
+guarantee to provide. Therefore, we should remove the second bullet point from the correctness property and instead add
+the following correctness property:
+> Any write from an upstream source that has been acknowledged by Materialize will be visible in all future queries.
+
+Proposal: When STORAGE wants to acknowledge some data that has a timestamp `ts`, it must wait until it knows that the
+global timestamp is larger than or equal to `ts`. STORAGE can discover this by one of the following ways (NOTE: these
+aren't separate proposals, they can all work together):
+
+- STORAGE will send an `ACK` request to the Coordinator timestamped with `ts`. The Coordinator will wait for the global
+  timestamp to advance to `ts` and then send a response back to STORAGE indicating that it's OK to send the
+  acknowledgement.
+- If STORAGE receives a read that doesn't use `AS OF`, is in the same timeline as the data being acknowledged, and has a
+  timestamp larger than or equal to `ts`, then it knows that the global timestamp must have advanced to `ts` or greater,
+  and it's safe to send the acknowledgement.
+- If STORAGE receives an updated Read Capability for the global timestamp (see
+  [Read Capabilities on Global Timestamp](#Read Capabilities on Global Timestamp)) larger than or equal to `ts`, then it
+  knows that the global timestamp must have advanced to `ts` or greater, and it's safe to send the acknowledgement.
+
+NOTE: There are some race conditions here such as:
+
+1. STORAGE sends an ACK request for `ts` to the Coordinator.
+2. STORAGE receives a read at `ts`.
+3. STORAGE sends the ACK for `ts` to an upstream source.
+4. STORAGES receives ACK OK response from the Coordinator.
+
+This is fine and STORAGE can just ignore the ACK OK response.
+
+## Alternatives
+
+- We don't hold read capabilities at the global timestamp. Instead, we block all queries that require a timestamp larger
+  than the global timestamp until the global timestamp advances to the timestamp of that
+  query. [#11150](https://github.com/MaterializeInc/materialize/pull/11150)
+  and [#10173](https://github.com/MaterializeInc/materialize/pull/10173) are previous pull requests that try and
+  implement this query blocking.
```

Comment:
is it vaguely the idea proposed here: https://github.com/MaterializeInc/materialize/pull/11262, but not global, timestamps at the granularity of sources; why did we move off of this towards batching?

Reply:
It is very similar to your proposal except slightly weaker. If i understand your proposal, all reads would get a timestamp larger than `upper` and block until `upper` advances. With this proposal, we still use the same logic to select timestamps as the current impl (which will result in a value less than `upper` a lot of the time).

I chose batching writes over this for the following reasons:
- My opinion is that blocking writes would be preferred to blocking reads (especially now considering that user tables are not heavily used).
- Batching writes allows us select and tune the exact amount of time that writes would block for while this could potentially result in unbounded blocking of reads.
- Group commit is well known/studied feature that many large systems use so I was more confident that it would be possible to implement without destroying perf.
- Batching writes also solves the problems discussed in #12198, so we get two for one.

Reply:
Though I think this could easily be updated to say always pick the highest upper of all objects and then block until the global timestamp catches up.